crates/crypto_api/src/crypto_system/mod.rs

```diff
@@ -100,6 +118,85 @@ pub trait CryptoSystem: Sync {
         message: &Box<dyn Buffer>,
         public_key: &Box<dyn Buffer>,
     ) -> CryptoResult<bool>;
+
+    // -- key exchange methods -- //
+
+    /// bytelength of key exchange seed
+    fn kx_seed_bytes(&self) -> usize;
+
+    /// bytelength of key exchange public key
+    fn kx_public_key_bytes(&self) -> usize;
+
+    /// bytelength of key exchange secret key
+    fn kx_secret_key_bytes(&self) -> usize;
+
+    /// bytelength of session keys derived from key exchange
+    fn kx_session_key_bytes(&self) -> usize;
+
+    /// generate a deterministic key exchange public / secret keypair
+    /// based off the given seed entropy
+    fn kx_seed_keypair(
+        &self,
+        seed: &Box<dyn Buffer>,
+        public_key: &mut Box<dyn Buffer>,
+        secret_key: &mut Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// generate a pure entropy based key exchange public / secret keypair
+    fn kx_keypair(
+        &self,
+        public_key: &mut Box<dyn Buffer>,
+        secret_key: &mut Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// generate key exchange session keys from "client" perspective
+    fn kx_client_session_keys(
+        &self,
+        client_rx: &mut Box<dyn Buffer>,
+        client_tx: &mut Box<dyn Buffer>,
+        client_pk: &Box<dyn Buffer>,
+        client_sk: &Box<dyn Buffer>,
+        server_pk: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// generate key exchange session keys from "server" perspective
+    fn kx_server_session_keys(
+        &self,
+        server_rx: &mut Box<dyn Buffer>,
+        server_tx: &mut Box<dyn Buffer>,
+        server_pk: &Box<dyn Buffer>,
+        server_sk: &Box<dyn Buffer>,
+        client_pk: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    // -- aead encryption methods -- //
+
+    /// bytelength of key exchange seed
+    fn aead_nonce_bytes(&self) -> usize;
+
+    /// bytelength of aead authentication tag
+    fn aead_auth_bytes(&self) -> usize;
+
+    /// bytelength of aead symmetric key
+    fn aead_secret_bytes(&self) -> usize;
+
+    fn aead_encrypt(
+        &self,
+        cipher: &mut Box<dyn Buffer>,
+        message: &Box<dyn Buffer>,
+        adata: Option<&Box<dyn Buffer>>,
+        nonce: &Box<dyn Buffer>,
+        secret: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
```

Comment:
Same as above comment- builder pattern or something (wrapper types? a struct with fields equal to the existing function args?) to make this less error prone

Reply:
named param macros!